rush-plugins/rush-http-build-cache-plugin/src/HttpBuildCacheProvider.ts

```diff
@@ -0,0 +1,395 @@
+        credentialsCache.deleteCacheEntry(this._credentialCacheId);
+        await credentialsCache.saveIfModifiedAsync();
+      }
+    );
+  }
+
+  private get _credentialCacheId(): string {
+    if (!this.__credentialCacheId) {
+      const cacheIdParts: string[] = [this._url.href];
+
+      if (this._isCacheWriteAllowedByConfiguration) {
+        cacheIdParts.push('cacheWriteAllowed');
+      }
+
+      this.__credentialCacheId = cacheIdParts.join('|');
+    }
+
+    return this.__credentialCacheId;
+  }
+
+  private async _http(options: {
+    terminal: ITerminal;
+    relUrl: string;
+    method: string;
+    body: BodyInit | undefined;
+    warningText: string;
+    readBody: boolean;
+    credentialOptions?: CredentialsOptions;
+  }): Promise<Buffer | boolean> {
+    const { terminal, relUrl, method, body, warningText, readBody, credentialOptions } = options;
+    const safeCredentialOptions = credentialOptions ?? CredentialsOptions.Optional;
+    const credentials = await this._tryGetCredentials(safeCredentialOptions);
+    const url = new URL(relUrl, this._url).href;
+
+    const headers: Record<string, string> = {};
+    if (typeof credentials === 'string') {
+      headers.Authorization = credentials;
+    }
+
+    for (const [key, value] of Object.entries(this._headers)) {
+      if (typeof value === 'string') {
+        headers[key] = value;
+      }
+    }
+
+    const bodyLength = (body as { length: number })?.length || 'unknown';
+
+    terminal.writeDebugLine(`[http-build-cache] request: ${method} ${url} ${bodyLength} bytes`);
+
+    const response = await fetch(url, {
+      method: method,
+      headers: headers,
+      body: body,
+      redirect: 'follow'
+    });
+
+    if (!response.ok) {
+      if (typeof credentials !== 'string' && safeCredentialOptions === CredentialsOptions.Optional) {
+        // We tried fetching the resource without credentials and that did not work out
+        // Try again but require credentials this time
+        // This will trigger the provider to request credentials
+        return await this._http({ ...options, credentialOptions: CredentialsOptions.Required });
+      }
+
+      this._reportFailure(terminal, method, response, false, warningText);
+      return false;
+    }
+
+    const result: Buffer | boolean = readBody ? Buffer.from(await response.arrayBuffer()) : true;
+
+    terminal.writeDebugLine(
+      `[http-build-cache] actual response: ${response.status} ${url} ${
+        result === true ? 'true' : result.length
+      } bytes`
+    );
+
+    return result;
+  }
+
+  private async _tryGetCredentials(options: CredentialsOptions.Required): Promise<string>;
+  private async _tryGetCredentials(options: CredentialsOptions.Optional): Promise<string | undefined>;
+  private async _tryGetCredentials(options: CredentialsOptions.Omit): Promise<undefined>;
+  private async _tryGetCredentials(options: CredentialsOptions): Promise<string | undefined>;
```

Comment:
```suggestion
```

Looks like this overload is redundant.

Reply:
I played with it, but it seems that TypeScript requires the full list of overloads if you have any overloads, otherwise you can't call the method with a generic CredentialsOptions object.